File: Leetcode/141/141.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    bool hasCycle(ListNode *head) {
        if (!head) {
            return false;
        }
        // Floyd check
        auto slow(head);
        auto fast(head);
        do {
            slow = slow->next;
            fast = fast->next;
            if (fast) {
                fast = fast->next;
            }
        } while (slow != fast && fast);
        return fast;
    }
};

File: Leetcode/141/141.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func hasCycle(head *ListNode) bool {
	if head == nil || head.Next == nil || head.Next.Next == nil {
		return false
	}
    // Floyd Cycle Detection
	slowNode := head.Next
	fastNode := head.Next.Next
	for slowNode != fastNode {
		if fastNode.Next == nil || fastNode.Next.Next == nil {
			return false
		}
		slowNode = slowNode.Next
		fastNode = fastNode.Next.Next
	}
	return true
}